src/topo.rs

```diff
@@ -0,0 +1,565 @@
+        let mut source_que = BTreeSet::new();
+        // This represents the candidates for relabeling to the left of
+        // the "line".
+        let mut source_stack = vec![];
+
+        // All the nodes that can be reached by our `to` node
+        // All nodes that have outgoing edges to `to`
+        // FromTarget in the paper
+        let mut target_que = BTreeSet::new();
+        // This represents the candidates for relabeling to the right of
+        // the line.
+        let mut target_stack = vec![];
+
+        // The number of incoming edges of source
+        let mut s_indegree = self
+            .nodes
+            .get(&from)
+            .ok_or(TopoError::NotFound)?
+            .in_degree();
+        // The number of outgoing edges of our target (to) node
+        let mut t_outdegree = self.nodes.get(&to).ok_or(TopoError::NotFound)?.out_degree();
+
+        // Each iteration `s` and `t` are our candidate nodes to be inserted in
+        // to their respected stacks
+        //
+        // `s` will be inserted into the `source_stack` when we find a node with
+        // equal incoming / outgoing edges
+        // TODO: why?
+        let mut s = Some(from);
+        // `t` will be inserted into the `target_stack` when we find a node with
+        // equal incoming / outgoing edges
+        // TODO: why?
+        let mut t = Some(to);
+
+        // Discover which nodes need relabeling...
+        // This is done by "placing a line" arbitrarily between `from` and `to`
+        while let (Some(some_s), Some(some_t)) = (s, t) {
+            // TODO: they ever going to let us `while let Some(foo) = bar if foo.is_something {...}`
+            if self.is_before(some_s, some_t)? {
+                break;
+            }
+
+            let d = s_indegree.min(t_outdegree);
+            s_indegree -= d;
+            t_outdegree -= d;
+
+            if s_indegree == 0 {
+                for idx_node_pair in self.nodes_with_edges_incoming(some_s) {
+                    source_que.insert(idx_node_pair);
+                }
+                source_stack.push(some_s);
+
+                s = if !source_que.is_empty() {
+                    // TODO:
+                    // use of unstable library feature 'map_first_last'
+                    // see issue #62924 <https://github.com/rust-lang/rust/issues/62924> for more information
+                    // `source_que.pop_last()`
+                    let tmp = source_que.iter().last().copied().unwrap();
+                    source_que.remove(&tmp);
+
+                    Some(tmp.1)
+                } else {
+                    // paper says "Then we are done"
+                    None
+                };
+
+                // We don't ever use this number again if `s` is None but
+                // we can't return early
+                s_indegree = s.map_or(0, |id| {
+                    self.nodes.get(&id).map(|n| n.in_degree()).unwrap_or(0)
+                });
```

Comment:
Not sure whether it's any better but this could be

```suggestion
                s_indegree = s
	                .and_then(|id| self.nodes.get(&id).map(|n| n.in_degree()))
	                .unwrap_or(0);
```

Reply:
Yeah, I think it reads better even if only because of the formatting. The double or 0 in my version is also confusing.